diff --git a/src/dispextern.h b/src/dispextern.h
index 213032d4de..f01e5fac47 100644
--- a/src/dispextern.h
+++ b/src/dispextern.h
@@ -2958,7 +2958,7 @@ reset_mouse_highlight (Mouse_HLInfo *hlinfo)
   void (*define_frame_cursor) (struct frame *f, Emacs_Cursor cursor);
 
   /* Clear the area at (X,Y,WIDTH,HEIGHT) of frame F.  */
-  void (*clear_frame_area) (struct frame *f, int x, int y,
+  void (*clear_frame_area) (struct window *w, struct frame *f, int x, int y,
                             int width, int height);
 
  /* Clear area of frame F's internal border.  If the internal border
@@ -3422,6 +3422,7 @@ #define TTY_CAP_STRIKE_THROUGH	0x20
 extern void handle_tool_bar_click (struct frame *,
                                    int, int, bool, int);
 
+extern bool expose_window (struct window *w, const Emacs_Rectangle *fr);
 extern void expose_frame (struct frame *, int, int, int, int);
 extern bool gui_intersect_rectangles (const Emacs_Rectangle *,
                                       const Emacs_Rectangle *,
@@ -3727,6 +3728,15 @@ #define IMAGE_BACKGROUND_TRANSPARENT(img, f, mask)			      \
 extern void gui_redo_mouse_highlight (Display_Info *);
 #endif /* HAVE_NS */
 
+extern bool isValidBackgroundImage ();
+#ifdef __OBJC__
+extern bool isBackgroundTransparentColor (NSColor *color);
+#endif
+extern bool setupBackgroundImageFile (const char * const filename);
+extern void drawBackground (NSRect rect);
+extern bool g_background_image_flag;
+extern float g_background_fill_alpha;
+
 #endif /* HAVE_WINDOW_SYSTEM */
 
 INLINE_HEADER_END
diff --git a/src/dispnew.c b/src/dispnew.c
index b3f7be67e0..479e28a5c5 100644
--- a/src/dispnew.c
+++ b/src/dispnew.c
@@ -4451,6 +4451,7 @@ scrolling_window (struct window *w, int tab_line_p)
   if (w32_use_visible_system_caret)
     return 0;
 #endif
+  return 0;
 
   /* Give up if some rows in the desired matrix are not enabled.  */
   if (! MATRIX_ROW_ENABLED_P (desired_matrix, i))
diff --git a/src/frame.c b/src/frame.c
index 4129a70aa5..f0d2d888c3 100644
--- a/src/frame.c
+++ b/src/frame.c
@@ -3919,6 +3919,9 @@ DEFUN ("frame-scale-factor", Fframe_scale_factor, Sframe_scale_factor,
   {"ns-appearance",		SYMBOL_INDEX (Qns_appearance)},
   {"ns-transparent-titlebar",	SYMBOL_INDEX (Qns_transparent_titlebar)},
 #endif
+  {"bg-image-enable-flag",	SYMBOL_INDEX (Qbg_image_enable_flag)},
+  {"bg-image-filename",		SYMBOL_INDEX (Qbg_image_filename)},
+  {"bg-fill-alpha",		SYMBOL_INDEX (Qbg_fill_alpha)},
 };
 
 #ifdef HAVE_WINDOW_SYSTEM
@@ -5047,6 +5050,32 @@ gui_set_no_special_glyphs (struct frame *f, Lisp_Object new_value, Lisp_Object o
     FRAME_NO_SPECIAL_GLYPHS (f) = !NILP (new_value);
 }
 
+void
+x_set_bg_image_enable_flag (struct frame *f, Lisp_Object arg, Lisp_Object oldval)
+{
+  g_background_image_flag = !NILP (arg);
+}
+
+void
+x_set_bg_image_filename (struct frame *f, Lisp_Object arg, Lisp_Object oldval)
+{
+  if (STRINGP(arg))
+    {
+      setupBackgroundImageFile (SDATA (arg));
+    }
+}
+
+void
+x_set_bg_fill_alpha (struct frame *f, Lisp_Object arg, Lisp_Object oldval)
+{
+  if (FLOATP(arg))
+    {
+      g_background_fill_alpha = XFLOAT_DATA (arg);
+    }
+  else
+    {
+    }
+}
 
 #ifndef HAVE_NS
 
@@ -6142,6 +6171,9 @@ syms_of_frame (void)
   DEFSYM (Qtop_only, "top-only");
   DEFSYM (Qiconify_top_level, "iconify-top-level");
   DEFSYM (Qmake_invisible, "make-invisible");
+  DEFSYM (Qbg_image_enable_flag, "bg-image-enable-flag");
+  DEFSYM (Qbg_image_filename, "bg-image-filename");
+  DEFSYM (Qbg_fill_alpha, "bg-fill-alpha");
 
   {
     int i;
diff --git a/src/frame.h b/src/frame.h
index 19ee6ac10e..aaee46b8d1 100644
--- a/src/frame.h
+++ b/src/frame.h
@@ -1637,6 +1637,10 @@ #define EMACS_CLASS "Emacs"
 extern void gui_set_alpha (struct frame *, Lisp_Object, Lisp_Object);
 extern void gui_set_no_special_glyphs (struct frame *, Lisp_Object, Lisp_Object);
 
+extern void x_set_bg_image_enable_flag (struct frame *, Lisp_Object, Lisp_Object);
+extern void x_set_bg_image_filename (struct frame *, Lisp_Object, Lisp_Object);
+extern void x_set_bg_fill_alpha (struct frame *, Lisp_Object, Lisp_Object);
+
 extern void validate_x_resource_name (void);
 
 extern Lisp_Object gui_display_get_resource (Display_Info *,
diff --git a/src/macfont.m b/src/macfont.m
index d86f09f485..bd877e414c 100644
--- a/src/macfont.m
+++ b/src/macfont.m
@@ -597,6 +597,7 @@ static void mac_font_get_glyphs_for_variants (CFDataRef, UTF32Char,
                            glyph_layouts, glyph_len, dir);
 }
 
+static bool get_cgcolor_is_transparent = false;
 static CGColorRef
 get_cgcolor(unsigned long idx, struct frame *f)
 {
@@ -607,6 +608,8 @@ static void mac_font_get_glyphs_for_variants (CFDataRef, UTF32Char,
   CGFloat *components = xmalloc (sizeof(CGFloat)*(1+noc));
   CGColorRef cgColor;
 
+  get_cgcolor_is_transparent = isBackgroundTransparentColor(nsColor);
+
   [nsColor getComponents: components];
   cgColor = CGColorCreate (colorSpace, components);
   xfree (components);
@@ -2914,8 +2917,26 @@ So we use CTFontDescriptorCreateMatchingFontDescriptor (no
           if (!face)
             face = FACE_FROM_ID (s->f, MOUSE_FACE_ID);
         }
-      CG_SET_FILL_COLOR_WITH_FACE_BACKGROUND (context, face, f);
-      CGContextFillRects (context, &background_rect, 1);
+      if (isValidBackgroundImage())
+        {
+          // CG_SET_FILL_COLOR_WITH_FACE_BACKGROUND (context, face, f);
+          {
+            CGColorRef refcol_ = get_cgcolor (NS_FACE_BACKGROUND (face), f);
+            CGColorRef refcol2_ = CGColorCreateCopyWithAlpha (refcol_, g_background_fill_alpha);
+            CGContextSetFillColorWithColor (context, refcol2_);
+            CGColorRelease (refcol_);
+            CGColorRelease (refcol2_);
+          }
+          drawBackground(NSRectFromCGRect(background_rect));
+          if (!get_cgcolor_is_transparent) {
+            CGContextFillRects (context, &background_rect, 1);
+          }
+        }
+      else
+        {
+          CG_SET_FILL_COLOR_WITH_FACE_BACKGROUND (context, face, f);
+          CGContextFillRects (context, &background_rect, 1);
+        }
     }
 
   if (macfont_info->cgfont)
diff --git a/src/nsfns.m b/src/nsfns.m
index 054777aa66..a21bc3d54b 100644
--- a/src/nsfns.m
+++ b/src/nsfns.m
@@ -967,6 +967,9 @@ Turn the input menu (an NSMenu) into a lisp list for tracking on lisp side.
   ns_set_appearance,
   ns_set_transparent_titlebar,
 #endif
+  x_set_bg_image_enable_flag,
+  x_set_bg_image_filename,
+  x_set_bg_fill_alpha,
 };
 
 
@@ -1384,6 +1387,13 @@ Turn the input menu (an NSMenu) into a lisp list for tracking on lisp side.
   gui_default_parameter (f, parms, Qfullscreen, Qnil,
                          "fullscreen", "Fullscreen", RES_TYPE_SYMBOL);
 
+  gui_default_parameter (f, parms, Qbg_image_enable_flag, Qnil,
+                         "bgImageEnableFlag", "BgImageEnableFlag", RES_TYPE_BOOLEAN);
+  gui_default_parameter (f, parms, Qbg_image_filename, Qnil,
+                         "bgImageFilename", "BgImageFilename", RES_TYPE_STRING);
+  gui_default_parameter (f, parms, Qbg_fill_alpha, Qnil,
+                         "bgFillAlpha", "BgFillAlpha", RES_TYPE_FLOAT);
+
   /* Allow set_window_size_hook, now.  */
   f->can_set_window_size = true;
 
diff --git a/src/nsterm.m b/src/nsterm.m
index b135e359f5..20ea9e7eb7 100644
--- a/src/nsterm.m
+++ b/src/nsterm.m
@@ -80,6 +80,40 @@ Updated by Christian Limpach (chris@nice.ch)
 #ifdef NS_IMPL_COCOA
 static EmacsMenu *mainMenu;
 #endif
+static NSImage *s_background_image = 0;
+static NSColor *s_background_transparent_color = 0;
+bool g_background_image_flag = false;
+float g_background_fill_alpha = 0.6f;
+bool isValidBackgroundImage ()
+{
+  return s_background_image && g_background_image_flag;
+}
+bool isBackgroundTransparentColor (NSColor *color)
+{
+  return (([color redComponent] == [s_background_transparent_color redComponent]) &&
+          ([color greenComponent] == [s_background_transparent_color greenComponent]) &&
+          ([color blueComponent] == [s_background_transparent_color blueComponent]));
+}
+bool setupBackgroundImageFile (const char * const filename)
+{
+  {
+    /* FIXME! 画像指定時の背景色が透明として扱われることに注意。(なので、あとから背景色を変えるとおかしくなる) */
+    struct frame *emacsframe = SELECTED_FRAME ();
+    s_background_transparent_color = ns_lookup_indexed_color (NS_FACE_BACKGROUND
+                                                              (FACE_FROM_ID (emacsframe, DEFAULT_FACE_ID)),
+                                                              emacsframe);
+  }
+  s_background_image = [[NSImage alloc] initWithContentsOfFile:[NSString stringWithCString:filename encoding:NSUTF8StringEncoding]];
+  return true;
+}
+void drawBackground (NSRect rect)
+{
+  NSGraphicsContext* theContext = [NSGraphicsContext currentContext];
+  [theContext saveGraphicsState];
+  [[NSColor colorWithPatternImage:s_background_image] set];
+  NSRectFill(rect);
+  [theContext restoreGraphicsState];
+}
 
 /* ==========================================================================
 
@@ -2911,7 +2945,14 @@ so some key presses (TAB) are swallowed by the system.  */
   ns_focus (f, &r, 1);
   [ns_lookup_indexed_color (NS_FACE_BACKGROUND
 			    (FACE_FROM_ID (f, DEFAULT_FACE_ID)), f) set];
-  NSRectFill (r);
+  if (isValidBackgroundImage())
+    {
+      drawBackground (r);
+    }
+  else
+    {
+      NSRectFill (r);
+    }
   ns_unfocus (f);
 
   /* as of 2006/11 or so this is now needed */
@@ -2923,7 +2964,7 @@ so some key presses (TAB) are swallowed by the system.  */
 
 
 static void
-ns_clear_frame_area (struct frame *f, int x, int y, int width, int height)
+ns_clear_frame_area (struct window *w, struct frame *f, int x, int y, int width, int height)
 /* --------------------------------------------------------------------------
     External (RIF):  Clear section of frame
    -------------------------------------------------------------------------- */
@@ -2941,7 +2982,14 @@ so some key presses (TAB) are swallowed by the system.  */
   ns_focus (f, &r, 1);
   [ns_lookup_indexed_color (NS_FACE_BACKGROUND (face), f) set];
 
-  NSRectFill (r);
+  if (isValidBackgroundImage())
+    {
+      drawBackground (r);
+    }
+  else
+    {
+      NSRectFill (r);
+    }
 
   ns_unfocus (f);
   return;
@@ -3001,10 +3049,23 @@ so some key presses (TAB) are swallowed by the system.  */
     NSRect dstRect = NSMakeRect (x, to_y, width, height);
     EmacsView *view = FRAME_NS_VIEW (f);
 
-    [view copyRect:srcRect to:dstRect];
+   if (isValidBackgroundImage())
+     {
+       Emacs_Rectangle r;
+       r.x = 0;
+       r.y = (from_y < to_y) ? from_y : to_y;
+       r.width = width;
+       r.height = height + abs(from_y - to_y);
+       w->must_be_updated_p = 1;
+       expose_window (w, &r);
+     }
+   else
+     {
+       [view copyRect:srcRect to:dstRect];
 #ifdef NS_IMPL_COCOA
-    [view setNeedsDisplayInRect:srcRect];
+       [view setNeedsDisplayInRect:srcRect];
 #endif
+     }
   }
 
   unblock_input ();
@@ -3104,8 +3165,8 @@ so some key presses (TAB) are swallowed by the system.  */
         }
       else
         {
-          ns_clear_frame_area (f, 0, y, width, height);
-          ns_clear_frame_area (f,
+          ns_clear_frame_area (w, f, 0, y, width, height);
+          ns_clear_frame_area (w, f,
                                FRAME_PIXEL_WIDTH (f) - width,
                                y, width, height);
         }
@@ -3397,7 +3458,14 @@ Note that CURSOR_WIDTH is meaningful only for (h)bar cursors.
   if (face && NS_FACE_BACKGROUND (face)
       == ns_index_color (FRAME_CURSOR_COLOR (f), f))
     {
-      [ns_lookup_indexed_color (NS_FACE_FOREGROUND (face), f) set];
+      if (isValidBackgroundImage())
+        {
+          [[ns_lookup_indexed_color (NS_FACE_FOREGROUND (face), f) colorWithAlphaComponent: g_background_fill_alpha] set];
+        }
+      else
+        {
+          [ns_lookup_indexed_color (NS_FACE_FOREGROUND (face), f) set];
+        }
       hollow_color = FRAME_CURSOR_COLOR (f);
     }
   else
@@ -3414,7 +3482,14 @@ Note that CURSOR_WIDTH is meaningful only for (h)bar cursors.
     case HOLLOW_BOX_CURSOR:
       NSRectFill (r);
       [hollow_color set];
-      NSRectFill (NSInsetRect (r, 1, 1));
+      if (isValidBackgroundImage())
+        {
+          drawBackground (NSInsetRect (r, 1, 1));
+        }
+      else
+        {
+          NSRectFill (NSInsetRect (r, 1, 1));
+        }
       [FRAME_CURSOR_COLOR (f) set];
       break;
     case HBAR_CURSOR:
@@ -3961,14 +4036,38 @@ Function modeled after x_draw_glyph_string_box ().
             }
           else
             face = FACE_FROM_ID (s->f, s->first_glyph->face_id);
+          bool is_bg_transparent = false;
           if (!face->stipple)
-            [(NS_FACE_BACKGROUND (face) != 0
-              ? ns_lookup_indexed_color (NS_FACE_BACKGROUND (face), s->f)
-              : FRAME_BACKGROUND_COLOR (s->f)) set];
+            {
+              if (isValidBackgroundImage())
+                {
+                  NSColor *bg_color = (NS_FACE_BACKGROUND (face) != 0
+                                       ? ns_lookup_indexed_color (NS_FACE_BACKGROUND (face), s->f)
+                                       : FRAME_BACKGROUND_COLOR (s->f));
+                  [[bg_color colorWithAlphaComponent: g_background_fill_alpha] set];
+                  if (isBackgroundTransparentColor(bg_color))
+                    {
+                      is_bg_transparent = true;
+                    }
+                }
+              else
+                {
+                  [(NS_FACE_BACKGROUND (face) != 0
+                    ? ns_lookup_indexed_color (NS_FACE_BACKGROUND (face), s->f)
+                    : FRAME_BACKGROUND_COLOR (s->f)) set];
+                }
+            }
           else
             {
               struct ns_display_info *dpyinfo = FRAME_DISPLAY_INFO (s->f);
-              [[dpyinfo->bitmaps[face->stipple-1].img stippleMask] set];
+              if (isValidBackgroundImage())
+                {
+                  [[[dpyinfo->bitmaps[face->stipple-1].img stippleMask] colorWithAlphaComponent: g_background_fill_alpha] set];
+                }
+              else
+                {
+                  [[dpyinfo->bitmaps[face->stipple-1].img stippleMask] set];
+                }
             }
 
           if (s->hl != DRAW_CURSOR)
@@ -3976,7 +4075,22 @@ Function modeled after x_draw_glyph_string_box ().
               NSRect r = NSMakeRect (s->x, s->y + box_line_width,
                                     s->background_width,
                                     s->height-2*box_line_width);
-              NSRectFill (r);
+              if (isValidBackgroundImage())
+                {
+                  if (is_bg_transparent)
+                    {
+                      drawBackground (r);
+                    }
+                  else
+                    {
+                      drawBackground (r);
+                      NSRectFillUsingOperation (r, NSCompositeSourceOver);
+                    }
+                }
+              else
+                {
+                  NSRectFill (r);
+                }
             }
 
 	  s->background_filled_p = 1;
@@ -4185,7 +4299,14 @@ Function modeled after x_draw_glyph_string_box ().
 
       glyphRect = NSMakeRect (s->x, s->y, s->background_width, s->height);
 
-      [bgCol set];
+      if (isValidBackgroundImage())
+        {
+          [[bgCol colorWithAlphaComponent: g_background_fill_alpha] set];
+        }
+      else
+        {
+          [bgCol set];
+        }
 
       /* NOTE: under NS this is NOT used to draw cursors, but we must avoid
          overwriting cursor (usually when cursor on a tab) */
@@ -4200,7 +4321,22 @@ overwriting cursor (usually when cursor on a tab) */
           glyphRect.size.width -= width;
           glyphRect.origin.x += width;
 
-          NSRectFill (glyphRect);
+          if (isValidBackgroundImage())
+            {
+              if (isBackgroundTransparentColor(bgCol))
+                {
+                  drawBackground (glyphRect);
+                }
+              else
+                {
+                  drawBackground (glyphRect);
+                  NSRectFillUsingOperation (glyphRect, NSCompositeSourceOver);
+                }
+            }
+          else
+            {
+              NSRectFill (glyphRect);
+            }
 
           /* Draw overlining, etc. on the cursor. */
           if (s->w->phys_cursor_type == FILLED_BOX_CURSOR)
@@ -4210,7 +4346,22 @@ overwriting cursor (usually when cursor on a tab) */
         }
       else
         {
+          if (isValidBackgroundImage())
+            {
+              if (isBackgroundTransparentColor(bgCol))
+                {
+                  drawBackground (glyphRect);
+                }
+              else
+                {
+                  drawBackground (glyphRect);
+                  NSRectFillUsingOperation (glyphRect, NSCompositeSourceOver);
+                }
+            }
+          else
+            {
           NSRectFill (glyphRect);
+            }
         }
 
       /* Draw overlining, etc. on the stretch glyph (or the part
@@ -4915,7 +5066,7 @@ in certain situations (rapid incoming events).
           [bar removeFromSuperview];
           wset_vertical_scroll_bar (window, Qnil);
           [bar release];
-          ns_clear_frame_area (f, left, top, width, height);
+          ns_clear_frame_area (window, f, left, top, width, height);
         }
       unblock_input ();
       return;
@@ -4924,7 +5075,7 @@ in certain situations (rapid incoming events).
   if (NILP (window->vertical_scroll_bar))
     {
       if (width > 0 && height > 0)
-	ns_clear_frame_area (f, left, top, width, height);
+        ns_clear_frame_area (window, f, left, top, width, height);
 
       bar = [[EmacsScroller alloc] initFrame: r window: win];
       wset_vertical_scroll_bar (window, make_mint_ptr (bar));
@@ -4939,7 +5090,7 @@ in certain situations (rapid incoming events).
       if (FRAME_LIVE_P (f) && !NSEqualRects (oldRect, r))
         {
           if (! NSEqualRects (oldRect, r))
-              ns_clear_frame_area (f, left, top, width, height);
+            ns_clear_frame_area (window, f, left, top, width, height);
           [bar setFrame: r];
         }
     }
@@ -5003,7 +5154,7 @@ in certain situations (rapid incoming events).
   if (NILP (window->horizontal_scroll_bar))
     {
       if (width > 0 && height > 0)
-	ns_clear_frame_area (f, left, top, width, height);
+        ns_clear_frame_area (window, f, left, top, width, height);
 
       bar = [[EmacsScroller alloc] initFrame: r window: win];
       wset_horizontal_scroll_bar (window, make_mint_ptr (bar));
@@ -5016,7 +5167,7 @@ in certain situations (rapid incoming events).
       oldRect = [bar frame];
       if (FRAME_LIVE_P (f) && !NSEqualRects (oldRect, r))
         {
-          ns_clear_frame_area (f, left, top, width, height);
+          ns_clear_frame_area (window, f, left, top, width, height);
           [bar setFrame: r];
           update_p = YES;
         }
@@ -5026,7 +5177,7 @@ in certain situations (rapid incoming events).
      a square that belongs to neither. We need to clear it otherwise
      it fills with junk.  */
   if (!NILP (window->vertical_scroll_bar))
-    ns_clear_frame_area (f, WINDOW_SCROLL_BAR_AREA_X (window), top,
+    ns_clear_frame_area (window, f, WINDOW_SCROLL_BAR_AREA_X (window), top,
                          NS_SCROLL_BAR_HEIGHT (f), height);
 
   if (update_p)
@@ -7603,6 +7754,7 @@ - (instancetype) initFrameFromEmacs: (struct frame *)f
   col = ns_lookup_indexed_color (NS_FACE_BACKGROUND
 				 (FACE_FROM_ID (emacsframe, DEFAULT_FACE_ID)),
 				 emacsframe);
+  s_background_transparent_color = col;
   [win setBackgroundColor: col];
   if ([col alphaComponent] != (EmacsCGFloat) 1.0)
     [win setOpaque: NO];
@@ -8528,7 +8680,7 @@ - (void)drawRect: (NSRect)rect
   int x = NSMinX (rect), y = NSMinY (rect);
   int width = NSWidth (rect), height = NSHeight (rect);
 
-  ns_clear_frame_area (emacsframe, x, y, width, height);
+  ns_clear_frame_area (0, emacsframe, x, y, width, height);
   block_input ();
   expose_frame (emacsframe, x, y, width, height);
   unblock_input ();
diff --git a/src/xdisp.c b/src/xdisp.c
index 23b4ba5c39..d26f5babf0 100644
--- a/src/xdisp.c
+++ b/src/xdisp.c
@@ -31226,7 +31226,7 @@ gui_clear_end_of_line (struct window *w, struct glyph_row *updated_row,
   if (to_x > from_x && to_y > from_y)
     {
       block_input ();
-      FRAME_RIF (f)->clear_frame_area (f, from_x, from_y,
+      FRAME_RIF (f)->clear_frame_area (w, f, from_x, from_y,
                                        to_x - from_x, to_y - from_y);
       unblock_input ();
     }
@@ -31752,7 +31752,7 @@ erase_phys_cursor (struct window *w)
       x = WINDOW_TEXT_TO_FRAME_PIXEL_X (w, x);
 
       if (width > 0)
-	FRAME_RIF (f)->clear_frame_area (f, x, y, width, cursor_row->visible_height);
+	FRAME_RIF (f)->clear_frame_area (w, f, x, y, width, cursor_row->visible_height);
     }
 
   /* Erase the cursor by redrawing the character underneath it.  */
@@ -34343,7 +34343,7 @@ gui_draw_bottom_divider (struct window *w)
    input blocked.  Value is true if the exposure overwrites
    mouse-face.  */
 
-static bool
+bool
 expose_window (struct window *w, const Emacs_Rectangle *fr)
 {
   struct frame *f = XFRAME (w->frame);
